spec/luassert_helper.lua

```diff
@@ -1,7 +1,20 @@
 local busted = require('busted')
 local assert = require('luassert.assert')
+local spy = require('luassert.spy')
+local stub = require('luassert.stub')
 local util = require ('luassert.util')
 local say = require('say')
+local tablex = require('pl.tablex')
+local deepcopy = tablex.deepcopy
+
+local original_stub_new = deepcopy(stub.new)
+
+stub.new = function(object, key, ...)
```

Comment:
Stub is a spy with extra functionality. So if we probably should fix the spy and that would also fix the stub.

`spy.is_spy(val)` retrurns true for both spies and stubs.

spy.new is called from stub.new, so just overriding spy.new should do the trick

Reply:
I've just done some test if I use `spy.new()` the callback argument is never spied, so I think that the main problem is here: 

https://github.com/Olivine-Labs/luassert/blob/3b2351c384cf982b953ab6d7964f835acb8cb7db/src/stub.lua#L27-L35

I'm not sure what is the best solution here, any opinion here?

Reply:
I'm not sure if I get it.

I'm just saying overriding `spy.new` instead of `stub.new` will cover both stubs and spies.

Reply:
This is an example:
 
```
localhost.localdomain ➜  APIcast git:(master) ✗ cat spec/stub_issue.lua
foo = {
    foo = function() return 1 end
}

describe("EloyIssue", function()
    it("Test Stubs", function()
        assert.same(foo.foo(), 1)

        stub(foo, "foo", function() return 2 end)
        assert.same(foo.foo(), 2)

        stub(foo, "foo", function() return 3 end)
        assert.same(foo.foo(), 3)
    end)
end)
localhost.localdomain ➜  APIcast git:(master) ✗ git diff spec/luassert_helper.lua | cat
diff --git a/spec/luassert_helper.lua b/spec/luassert_helper.lua
index 8cd07802..a2e5c8ef 100644
--- a/spec/luassert_helper.lua
+++ b/spec/luassert_helper.lua
@@ -1,8 +1,16 @@
 local busted = require('busted')
 local assert = require('luassert.assert')
 local util = require ('luassert.util')
+local spy = require ('luassert.spy')
 local say = require('say')

+original_spy_new = spy.new
+
+function spy.new(callback)
+    print("Callback is spy:", spy.is_spy(callback))
+    return original_spy_new(callback)
+end
+
 do -- set up reverting stubs
   local state = require("luassert.state")

```

```
bash-4.2$
bash-4.2$ clear
bash-4.2$ BUSTED_FILES=spec/stub_issue.lua make busted
/usr/local/openresty/luajit/bin/rover install --roverfile=gateway/Roverfile > /dev/null
/usr/local/openresty/luajit/bin/rover exec bin/busted spec/stub_issue.lua
Callback is spy:false
Callback is spy:false
Callback is spy:false
●
1 success / 0 failures / 0 errors / 0 pending : 0.003327 seconds
bash-4.2$
```
For some reason, the spy.new the callback always is never spied, I also tried on spy.on but in the only place that works is on `stub.new()`

Reply:
I see. `spy.new` has a different signature from `stub.new`. Then stub `spy.on` ?

It is being called from `stub.new`. https://github.com/Olivine-Labs/luassert/blob/3b2351c384cf982b953ab6d7964f835acb8cb7db/src/stub.lua#L38